cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,29 +466,63 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
+    File archive = new File(tempDir, archiveName);
+    try (
+      ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+      zipOutput.setLevel(Deflater.NO_COMPRESSION);
+      for (File file : getLocalSparkLibrary(sparkCompat)) {
+        zipOutput.putNextEntry(new ZipEntry(file.getName()));
+        Files.copy(file.toPath(), zipOutput);
+        zipOutput.closeEntry();
+      }
+    }
 
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
+    Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+    //Create writing file
+    boolean isWriteComplete = false;
+    while (!isWriteComplete) {
+      if (writeInProgressLoc.createNew()) {
+        Location tempDirHdfs = locationFactory.create("/framework/spark_temp_" + UUID.randomUUID());
+        Location tempArchiveHdfs = tempDirHdfs.append(archiveName);
```

Comment:
You can use `frameworkLocation.getTempFile` instead. No need to create a temp directory.